README.md

```diff
@@ -152,69 +151,39 @@ Upon a call to `wasi_thread_spawn`, the WASI host must:
-
-  * The most significant bit is the sign bit. As `wasi_thread_spawn` uses
-    signed integer and uses negative values to indicate errors, a TID needs
-    to be positive.
-
-  * The second bit need to be 0 in order to be compatible with the TID-based
-    locking implementation in musl/wasi-libc.
-
-  * The third bit need to be 0 in order to make an extra room for other
-    reserved values in wasi-libc.
-    For example, it can be used to indicate the main thread, which doesn't
-    have a TID in the current version of this proposal.
-
-### Process
-
-* A process is a group of threads.
-
-* The main thread starts with a process which only contains
-  the main thread.
-
-* Threads created by a thread in a process using `wasi_thread_spawn`
-  are added to the process.
-
-* When a thread is terminated, it's removed from the process.
-
-### Voluntary thread termination
-
-A thread can terminate itself voluntarily by returning from
-`wasi_thread_start`.
-
-### Changes to WASI `proc_exit`
-
-With this proposal, the `proc_exit` function takes extra responsibility
-to terminate all threads in the process, not only the calling one.
-
-Any of the threads in the process can call `proc_exit`.
-
-### Traps
-
-When a trap occurs in any thread, the entire process is terminated.
-
-### Process exit status
-
-If one or more threads call WASI `proc_exit` or raise a trap,
-one of them is chosen by the runtime to represent the exit status
-of the process.
-It's non deterministic which one is chosen.
-
-If all the threads in the process have been terminated without calling
-`proc_exit` or raising a trap, it's treated as if the last thread called
-`proc_exit` with exit code 0.
+#### Design choice: thread IDs
+
+When `wasi_thread_spawn` successfully spawns a thread, it returns a thread ID
+(TID) &mdash; 32-bit integer with several restrictions. TIDs are managed and
+provided by the WASI host. To avoid compatibility issues in wasi-libc, `0` is
+reserved; `wasi_thread_spawn` should not return it. To avoid leaking
+information, the host may choose to return random TIDs (as opposed to leaking OS
+TIDs).
+
+TIDs returned by `wasi_thread_spawn` must always zero the uppermost three bits:
+- the most significant bit is the sign bit; `wasi_thread_spawn` uses negative
+  values to indicate errors, so a valid TID must zero this bit (i.e., a valid
+  TID is always positive)
+- the second most significant bit is used for TID-based locking in wasi-libc
+  &mdash; this must be zeroed initially in a valid TID
+- the third most significant bit allows wasi-libc to mark threads as the main
+  thread; it is zeroed by any thread returned by `wasi_thread_spawn`
+
+#### Design choice: termination
+
+A `wasi-threads` module initially executes a single thread &mdash; the main
+thread. As `wasi_thread_spawn` is called, more threads begin to execute. Threads
+terminate in the following ways:
+
+- __voluntarily__, by returning from `wasi_thread_start`; other threads continue
```

Comment:
How about `__normally__` or `__naturally__`?

Reply:
or maybe just remove `__voluntarily__`?

after all, all of these reasons are in some senses normal, natural, and voluntary.

Reply:
I somewhat disagree.  I think it makes sense to refer to natural/normal exiting as opposed to premature/exceptional exiting where control flow doesn't flow naturally out of the entry point.

Reply:
while i can understand what you mean, i still feel it's confusing as even wasi is saying `proc_exit` teminates "normally".
https://github.com/WebAssembly/WASI/blob/069e3a4aee7ecaa8b52bd3c0ebe4fc055c1b1951/phases/snapshot/witx/wasi_snapshot_preview1.witx#L440

Reply:
How about "upon finishing execution"? Then we have the symmetry with "upon a trap" and "upon a `proc_exit` call".

Reply:
"Upon returning from the entry point"?